src/main/webapp/app/exercises/programming/participate/programming-submission.service.ts

```diff
@@ -217,6 +303,78 @@ export class ProgrammingSubmissionService implements IProgrammingSubmissionServi
+                newSubmissionTopic = this.SUBMISSION_PROCESSING_TEMPLATE_TOPIC.replace('%exerciseId%', exerciseId.toString());
+            }
+
+            // Only subscribe if not subscription to same topic exists (e.g. from different participation)
+            if (!Array.from(this.submissionProcessingTopicsSubscribed.values()).includes(newSubmissionTopic)) {
+                this.websocketService.subscribe(newSubmissionTopic);
+                this.websocketService
+                    .receive(newSubmissionTopic)
+                    .pipe(
+                        tap((submissionProcessing: SubmissionProcessingDTO) => {
+                            const submissionParticipationId = submissionProcessing.participationId!;
+                            const exerciseId = this.participationIdToExerciseId.get(submissionParticipationId)!;
+
+                            if (!this.isNewestSubmission(submissionProcessing, exerciseId, submissionParticipationId)) {
+                                return;
+                            }
+
+                            const programmingSubmission = this.getSubmissionByCommitHash(submissionProcessing);
+                            // It is possible that the submission started processing before it got saved to the database and the message was sent to the client.
+                            // In this case, we cache that the submission started processing and do not emit the building state.
+                            // When the submission message arrives, we check if the submission is already in the cache.
+                            if (!programmingSubmission) {
+                                this.startedProcessingCache.set(submissionProcessing.commitHash!, {
+                                    estimatedCompletionDate: submissionProcessing.estimatedCompletionDate,
+                                    buildStartDate: submissionProcessing.buildStartDate,
+                                });
+                                return;
+                            }
+                            programmingSubmission.isProcessing = true;
+
+                            const buildTimingInfo = {
+                                estimatedCompletionDate: submissionProcessing.estimatedCompletionDate,
+                                buildStartDate: submissionProcessing.buildStartDate,
+                            };
+                            this.removeSubmissionFromProcessingCache(programmingSubmission.commitHash!);
+                            this.resetQueueEstimateTimer(submissionParticipationId);
+                            this.emitBuildingSubmission(submissionParticipationId, exerciseId, programmingSubmission, buildTimingInfo);
+                            // Now we start a timer, if there is no result when the timer runs out, it will notify the subscribers that no result was received and show an error.
+                            this.startResultWaitingTimer(submissionParticipationId);
```

Comment:
I do not think that it is good that we have this comment duplicated - would it make more sense as javadoc for the function?